README.md

```diff
@@ -1,16 +1,715 @@
+
+The invocation wrapper MUST be signed by the same principal that issued the UCAN.
+
+## 3.1 Fields
+
+| Field | Type                     | Description                                            | Required | Default |
+|-------|--------------------------|--------------------------------------------------------|----------|---------|
+| `v`   | `SemVer`                 | SemVer of the UCAN invocation this schema (v0.1.0)     | Yes      |         |
+| `prf` | `[&UCAN]`                | UCANs that supply the authority to run this invocation | Yes      |         |
+| `run` | `"*" or {String : Task}` | Which UCAN capabilities to run                         | Yes      |         |
+| `nnc` | `String`                 | A unique nonce, to distinguish each invocation         | Yes      |         |
+| `ext` | `Any`                    | Non-normative extended fields                          | No       | `null`  |
+    
+### 3.1.1 Version
+
+The `v` field MUST contain the version of the invocation object  schema.
+
+### 3.1.2 Proofs
+
+The `prf` field MUST contain CIDs pointing to the UCANs that provide the authority to run these tasks. The elements of this array MUST be sorted in ascending numeric order. Restricting the outmost UCANs to only the authority required for the current invocation is RECOMMENDED.
+
+### 3.1.3 Run Capabilities
+
+The `run` field MUST reference the tasks contained in the UCAN are to be run during the invocation. To run all tasks in the underlying UCAN, the `"*"` value MUST be used. If only specific tasks (or [pipelines](#5-promise-pipelining)) are intended to be run, then they MUST be prefixed with an arbitrary label and treated as a UCAN attenuation: all tasks MUST be backed by a matching capability of equal or greater authority.
+
+#### 3.1.3.1 Promises
+
+The only difference from general capabilities is that [promises](#5-promise-pipelining) MAY also be used as inputs to attenuated fields.
+
+If a capability input has the key `"_"` and the value is a promise, the input MUST be discarded and only used for determining sequencing tasks.
+
+### 3.1.4 Nonce
+
+The `nnc` field MUST contain a unique nonce. This field exists to enable making the CID of each invocation unique. While this field MAY be an empty string, it is NOT RECOMMENDED.
+
+### 3.1.5 Extended Fields
+
+The OPTIONAL `ext` field MAY contain arbitrary data. If not present, the `ext` field MUST be interpreted as `null`, including for [signature](#315-signature).
+
+## 3.2 IPLD Schema
+
+``` ipldsch
+type SignedInvocation struct {
+  inv Invocation (rename "ucan/invoke") 
+  sig VarSig
+}
+
+type Invocation struct {
+  v   SemVer  -- Version
+  prf [&UCAN] -- Authority to run this invocation
+  run Scope   -- Which tasks to invoke
+  nnc String  -- Nonce
+  ext nullable Any (implicit null) -- Extended fields
+}
+
+type Scope union {
+  | All "*"
+  | {String : Task}
+}
+
+type Task struct {
+  with   URI 
+  do     Ability
+  inputs Any
+  meta   {String : Any}
+} 
```

Comment:
So in this doc we sort of just mention `Task` in passing. In the [IPVM doc](https://github.com/ipvm-wg/workflow/pull/1/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R281) we treat them as a first class citizen (and they even get a quote!)

If th IPVM spec is going to reference this one, it seems like having most of that detail in here would be wise. And then the IPVM spec can just cover how IPVM workflows use them, e.g. the keys they expect to see in `meta`.

I don't really know where the Docker and/or WASM standardisation goes yet (e.g. just the fact that in the Docker type the `with` field should reference an image).

Reply:
Another important thing is talking about canonicalisation somewhere e.g. what does "canonical" mean for the `with` and `inputs` fields.

Reply:
I think maybe the above comments were from before our call yesterday, but can't hurt to repeat here for others lurking 👀 

> I don't really know where the Docker and/or WASM standardisation goes yet (e.g. just the fact that in the Docker type the with field should reference an image).

There are very roughly two kinds of spec:

1. Maxi: A manual for how to implement a complete end-to-end system
2. Mini: The minimal number of things everyone agrees on to be interoperable

Maxi specs are generally written internally as a project bible. Mini specs tend to get used to align many orgs who have overlapping-but-differing aims. The balance is in not making a mini spec [_so general_ as to be meaningless](https://en.wikipedia.org/wiki/OAuth#Controversy).

My aim for this work is to keep it in the mini camp, and empower others to do things that I never expected with it. If someone runs off and does something novel and awesome, then we declare success 😉

The last thing that I want is for everyone to have to register their stuff into a central registry. Open source standards have been moving away from that style for the past while, towards open namespacing and open interoperability.

To keep things openly interoperable, standardization of Docker and Wasm (and other!) fields would live in their own specs. If they're in this spec, then anyone who wants to conform to this spec would need to support those fields, too. Updating those fields would require a potentially major version bump to the entire spec, and would be a lot less agile.

The way I'm imagining it right now (and would love to know if there's a better way) is that we define `wasm-task` and `docker-task` specs. For IPVM, we'll also need to define `https-task`, and so on. If someone wants to bring RISC-V or CUDA tasks, then they can write their own specs, and at least the IPVM service discovery layer will disseminate this info and help the scheduler know how to route tasks to.

Reply:
I'll write one here as an example: https://github.com/ipvm-wg/task-wasm

(Blank at time of writing, but may have some text later today)